File: sync.js
/**
 * Garmin data sync for running tracker.
 * Fetches activities, VO2Max, weight, sleep for each user and merges into data/*.json
 *
 * Usage:
 *   node sync.js              → sync all users
 *   node sync.js --user yunho → sync only yunho
 *   node sync.js --user gf    → sync only girlfriend
 */

import { readFileSync, writeFileSync, existsSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { homedir } from "node:os";

const __dirname = dirname(fileURLToPath(import.meta.url));

const USERS = {
  yunho: {
    sessionFile: join(__dirname, "sessions/yunho-session.json"),
    dataFile: join(__dirname, "data/yunho.json"),
    name: "윤호",
    birthYear: 1992,
    targetWeightKg: 70,
    goalDistanceKm: 21.1,
    goalTimeMin: 120,
  },
  gf: {
    sessionFile: join(__dirname, "sessions/gf-session.json"),
    dataFile: join(__dirname, "data/gf.json"),
    name: "Jenny",
    birthYear: 2001,
    targetWeightKg: null,
    goalDistanceKm: 20,
    goalTimeMin: null,
    maxRunsPerWeek: 3,
    allEasy: true,            // no intensity work
    hrCeiling: 155,           // MAF — flag if avg HR exceeds this
    targetPaceSecMin: 470,    // 7:50/km
    targetPaceSecMax: 490,    // 8:10/km
    longRunMaxIncreasePct: 10,
    longRunFlagPct: 15,
    goalNote: "2026년 말까지 20km 편안하게 완주 · 주 3회 올 이지",
  },
};

function isJwtExpiredOrSoon(jwtValue, bufferSec = 3600) {
  try {
    const payload = JSON.parse(Buffer.from(jwtValue.split(".")[1], "base64url").toString());
    return payload.exp * 1000 < Date.now() + bufferSec * 1000;
  } catch { return true; }
}

async function refreshSessionIfNeeded(config) {
  const session = JSON.parse(readFileSync(config.sessionFile, "utf-8"));
  const jwtCookie = session.cookies.find((c) => c.name === "JWT_WEB");
  const needsRefresh = !session.csrf_token || (jwtCookie && isJwtExpiredOrSoon(jwtCookie.value));
  if (!needsRefresh) return;

  console.log(`[${config.name}] 세션 갱신 중 (CSRF=${!session.csrf_token ? "없음" : "만료"})...`);
  let playwright;
  try { playwright = await import("playwright"); }
  catch { throw new Error("playwright not available"); }

  const browser = await playwright.chromium.launch({ headless: true });
  const context = await browser.newContext();
  await context.addCookies(session.cookies.map((c) => ({
    name: c.name, value: c.value, domain: c.domain, path: "/",
  })));

  const page = await context.newPage();
  try {
    // Navigate to app — triggers JWT refresh via SSO if session cookie is valid
    // Also extracts CSRF token from the page meta tag
    await page.goto("https://connect.garmin.com/app/dashboard", {
      waitUntil: "domcontentloaded",
      timeout: 30000,
    });
    // Brief wait for any post-load auth redirects to settle
    await page.waitForTimeout(3000);
    const csrf = await page.evaluate(
      "document.querySelector('meta[name=\"csrf-token\"]')?.content ?? null"
    );
    const newCookies = await context.cookies();
    const garminCookies = newCookies
      .filter((c) => c.domain?.includes("garmin"))
      .map((c) => ({ name: c.name, value: c.value, domain: c.domain }));

    const newJwt = garminCookies.find((c) => c.name === "JWT_WEB");
    if (newJwt && !isJwtExpiredOrSoon(newJwt.value, 60)) {
      session.cookies = garminCookies;
      if (csrf) session.csrf_token = csrf;
      writeFileSync(config.sessionFile, JSON.stringify(session, null, 2));
      if (config.sessionFile.includes("yunho")) {
        writeFileSync(join(homedir(), ".garmin-connect-mcp/session.json"),
          JSON.stringify(session, null, 2));
      }
      console.log(`[${config.name}] 세션 갱신 완료 ✓ (CSRF: ${csrf ? "획득" : "없음"})`);
    } else {
      console.log(`[${config.name}] 세션 갱신 실패 — 기존으로 계속 시도`);
    }
  } finally {
    await browser.close();
  }
}

async function fetchUserData(userId, config) {
  if (!existsSync(config.sessionFile)) {
    console.log(`[${config.name}] 세션 파일 없음: ${config.sessionFile} — 건너뜀`);
    return null;
  }

  await refreshSessionIfNeeded(config);

  let GarminClient;
  try {
    const mod = await import("@etweisberg/garmin-connect-mcp");
    GarminClient = mod.GarminClient;
  } catch {
    // Try local node_modules path
    const localMod = await import(join(__dirname, "node_modules/@etweisberg/garmin-connect-mcp/dist/garmin-client.js"));
    GarminClient = localMod.GarminClient;
  }

  const client = new GarminClient(config.sessionFile);

  try {
    console.log(`[${config.name}] 데이터 수집 시작...`);

    // Fetch profile to get displayName
    const profile = await client.get("userprofile-service/userprofile/settings");
    const displayName = profile.displayName;
    console.log(`[${config.name}] 프로필: ${displayName}`);

    // Fetch last 50 running activities
    const activitiesRes = await client.get("activitylist-service/activities/search/activities", {
      start: 0,
      limit: 50,
      activityType: "running",
    });
    const activities = (activitiesRes || []).map((a) => ({
      id: a.activityId,
      date: a.startTimeLocal?.substring(0, 10) ?? "",
      distanceM: Math.round(a.distance ?? 0),
      durationSec: Math.round(a.duration ?? 0),
      avgHR: a.averageHR ?? null,
      maxHR: a.maxHR ?? null,
      avgPaceSecPerKm: a.distance > 0 ? Math.round((a.duration / (a.distance / 1000))) : null,
      calories: a.calories ?? null,
      name: a.activityName ?? "",
    }));
    console.log(`[${config.name}] 러닝 ${activities.length}개 수집`);

    // Fetch VO2Max
    let vo2max = null;
    try {
      const today = new Date().toISOString().split("T")[0];
      const vo2Res = await client.get(`metrics-service/metrics/maxmet/latest/${today}`);
      if (vo2Res?.generic?.calendarDate) {
        vo2max = [{ date: vo2Res.generic.calendarDate, value: vo2Res.generic.vo2MaxPreciseValue ?? vo2Res.generic.vo2MaxValue }];
      }
    } catch (e) {
      console.log(`[${config.name}] VO2Max 수집 실패: ${e.message}`);
    }

    // Fetch weight (last 12 months)
    let weight = null;
    try {
      const endDate = new Date().toISOString().split("T")[0];
      const startDate = new Date(Date.now() - 365 * 86400000).toISOString().split("T")[0];
      const weightRes = await client.get(`weight-service/weight/range/${startDate}/${endDate}`, { includeAll: "true" });
      if (weightRes?.dateWeightList) {
        weight = weightRes.dateWeightList
          .map((w) => ({
            date: w.calendarDate,
            kg: w.weight ? w.weight / 1000 : null, // Garmin stores in grams
          }))
          .filter((w) => w.kg !== null);
      }
    } catch (e) {
      console.log(`[${config.name}] 체중 수집 실패: ${e.message}`);
    }

    // Fetch today's sleep score
    let sleep = null;
    try {
      const today = new Date().toISOString().split("T")[0];
      const sleepRes = await client.get("sleep-service/sleep/dailySleepData", {
        date: today,
        nonSleepBufferMinutes: 60,
      });
      if (sleepRes?.dailySleepDTO) {
        sleep = [{
          date: sleepRes.dailySleepDTO.calendarDate,
          score: sleepRes.dailySleepDTO.sleepScores?.overall?.value ?? null,
          durationMin: sleepRes.dailySleepDTO.sleepTimeSeconds
            ? Math.round(sleepRes.dailySleepDTO.sleepTimeSeconds / 60)
            : null,
        }];
      }
    } catch (e) {
      console.log(`[${config.name}] 수면 수집 실패: ${e.message}`);
    }

    // Fetch training readiness (today)
    let trainingReadiness = null;
    try {
      const today = new Date().toISOString().split("T")[0];
      const trRes = await client.get(`metrics-service/metrics/trainingreadiness/${today}`);
      const tr = Array.isArray(trRes) ? trRes[0] : trRes;
      if (tr?.score != null) {
        trainingReadiness = {
          date: tr.calendarDate ?? today,
          score: tr.score,
          level: tr.level ?? null,
          feedbackShort: tr.feedbackShort ?? null,
          sleepScore: tr.sleepScore ?? null,
          recoveryTimeMin: tr.recoveryTime ?? null,
        };
      }
    } catch (e) {
      console.log(`[${config.name}] 훈련 준비도 수집 실패: ${e.message}`);
    }

    await client.close();

    const age = config.birthYear ? new Date().getFullYear() - config.birthYear : null;
    return {
      userId,
      name: config.name,
      garminDisplayName: displayName,
      birthYear: config.birthYear,
      age,
      mafHR: age ? 180 - age : null,
      targetWeightKg: config.targetWeightKg,
      goalDistanceKm: config.goalDistanceKm,
      goalTimeMin: config.goalTimeMin,
      activities,
      vo2max,
      weight,
      sleep,
      trainingReadiness,
      lastSync: new Date().toISOString(),
    };
  } catch (err) {
    await client.close().catch(() => {});
    throw err;
  }
}

function mergeData(existing, fresh) {
  if (!existing) return fresh;

  // Merge activities: keep all existing + add new ones (by id)
  const existingIds = new Set((existing.activities || []).map((a) => a.id));
  const newActivities = (fresh.activities || []).filter((a) => !existingIds.has(a.id));
  const mergedActivities = [...newActivities, ...(existing.activities || [])]
    .sort((a, b) => b.date.localeCompare(a.date))
    .slice(0, 200); // keep last 200

  // Merge weight: deduplicate by date, keep all
  const existingWeightDates = new Set((existing.weight || []).map((w) => w.date));
  const newWeight = (fresh.weight || []).filter((w) => !existingWeightDates.has(w.date));
  const mergedWeight = [...(existing.weight || []), ...newWeight].sort((a, b) => b.date.localeCompare(a.date));

  // Merge VO2Max: same
  const existingV2Dates = new Set((existing.vo2max || []).map((v) => v.date));
  const newV2 = (fresh.vo2max || []).filter((v) => !existingV2Dates.has(v.date));
  const mergedV2 = [...(existing.vo2max || []), ...newV2].sort((a, b) => b.date.localeCompare(a.date));

  return {
    ...existing,
    ...fresh,
    activities: mergedActivities,
    weight: mergedWeight,
    vo2max: mergedV2,
  };
}

async function syncUser(userId) {
  const config = USERS[userId];
  let existing = null;

  if (existsSync(config.dataFile)) {
    try {
      existing = JSON.parse(readFileSync(config.dataFile, "utf-8"));
    } catch {}
  }

  const fresh = await fetchUserData(userId, config);
  if (!fresh) return;

  const merged = mergeData(existing, fresh);
  writeFileSync(config.dataFile, JSON.stringify(merged, null, 2));
  console.log(`[${config.name}] 저장 완료 → ${config.dataFile}`);

  // Also sync session back to ~/.garmin-connect-mcp/session.json for MCP use
  if (userId === "yunho") {
    writeFileSync(`${process.env.HOME}/.garmin-connect-mcp/session.json`,
      readFileSync(config.sessionFile));
  }
}

async function main() {
  const args = process.argv.slice(2);
  const userArg = args.find((a) => a.startsWith("--user="))?.split("=")[1]
    ?? (args.indexOf("--user") !== -1 ? args[args.indexOf("--user") + 1] : null);

  const usersToSync = userArg ? [userArg] : Object.keys(USERS);

  for (const userId of usersToSync) {
    if (!USERS[userId]) {
      console.error(`알 수 없는 유저: ${userId}`);
      continue;
    }
    try {
      await syncUser(userId);
    } catch (err) {
      console.error(`[${USERS[userId].name}] 동기화 실패: ${err.message}`);
      if (err.message.includes("401")) {
        console.error(`  → 세션 만료됨. ${userId === "yunho" ? "윤호" : "Jenny"} 가민에 재로그인 후 쿠키를 sessions/${userId}-session.json에 업데이트해주세요.`);
      }
    }
  }
}

async function generateDashboard() {
  try {
    const { default: generate } = await import("./generate-dashboard.js?" + Date.now());
  } catch (e) {
    const { execSync } = await import("node:child_process");
    execSync("node generate-dashboard.js", { cwd: __dirname, stdio: "inherit" });
  }
}

async function pushDashboard() {
  const { execSync } = await import("node:child_process");
  try {
    execSync('git add dashboard.html', { cwd: __dirname, stdio: "pipe" });
    const today = new Date().toISOString().substring(0, 10);
    execSync(`git commit -m "dashboard: ${today} 자동 갱신"`, { cwd: __dirname, stdio: "pipe" });
    execSync('git push', { cwd: __dirname, stdio: "inherit" });
    console.log("대시보드 GitHub 푸시 완료 ✓");
  } catch (e) {
    // No changes to commit, or push failed — non-fatal
    if (!e.message?.includes("nothing to commit")) {
      console.log("GitHub 푸시 실패 (원격 연결 확인):", e.message?.split("\n")[0]);
    }
  }
}

main()
  .then(() => generateDashboard())
  .then(() => pushDashboard())
  .catch(console.error);

File: sync.py
"""
Garmin data sync — uses garminconnect (Python) for auth + data fetching.
Saves to data/yunho.json and data/gf.json in same format as before.

Usage:
  python3 sync.py              → sync all users
  python3 sync.py --user yunho → sync only yunho
  python3 sync.py --user gf    → sync only gf
"""

import os, sys, json, datetime
from pathlib import Path

BASE = Path(__file__).parent

# Load .env
env_path = BASE / ".env"
if env_path.exists():
    for line in env_path.read_text().splitlines():
        if "=" in line:
            k, _, v = line.partition("=")
            os.environ.setdefault(k.strip(), v.strip())

USERS = {
    "yunho": {
        "data_file": BASE / "data/yunho.json",
        "token_dir": str(BASE / "sessions/garth-yunho"),
        "email_env": "GARMIN_EMAIL",
        "password_env": "GARMIN_PASSWORD",
        "name": "윤호",
        "birth_year": 1992,
        "target_weight_kg": 70,
        "goal_distance_km": 21.1,
        "goal_time_min": 120,
    },
    "gf": {
        "data_file": BASE / "data/gf.json",
        "token_dir": str(BASE / "sessions/garth-gf"),
        "email_env": "GARMIN_GF_EMAIL",
        "password_env": "GARMIN_GF_PASSWORD",
        "name": "Jenny",
        "birth_year": 2001,
        "target_weight_kg": None,
        "goal_distance_km": 20,
        "goal_time_min": None,
        "max_runs_per_week": 3,
        "all_easy": True,
        "hr_ceiling": 155,
        "target_pace_sec_min": 470,
        "target_pace_sec_max": 490,
        "long_run_flag_pct": 15,
        "goal_note": "2026년 말까지 20km 편안하게 완주 · 주 3회 올 이지",
    },
}


def get_client(cfg):
    from garminconnect import Garmin

    token_dir = cfg["token_dir"]
    email = os.environ.get(cfg["email_env"])
    password = os.environ.get(cfg["password_env"])

    client = Garmin(email, password)
    try:
        client.login(tokenstore=token_dir)
        print(f"[{cfg['name']}] 저장된 토큰으로 로그인 성공")
    except Exception:
        print(f"[{cfg['name']}] 토큰 없음 — 새로 로그인 중...")
        if not email or not password:
            raise RuntimeError(f"{cfg['email_env']} / {cfg['password_env']} 미설정")
        client.login()
        client.garth.dump(token_dir)
        print(f"[{cfg['name']}] 로그인 완료, 토큰 저장")
    return client


def fetch_activities(client, cfg, start=0, limit=50):
    try:
        raw = client.get_activities(start, limit)
        result = []
        for a in raw:
            if a.get("activityType", {}).get("typeKey") != "running":
                continue
            dist = a.get("distance", 0) or 0
            dur = a.get("duration", 0) or 0
            pace = round(dur / (dist / 1000)) if dist > 0 else None
            result.append({
                "id": a.get("activityId"),
                "date": (a.get("startTimeLocal") or "")[:10],
                "distanceM": round(dist),
                "durationSec": round(dur),
                "avgHR": a.get("averageHR"),
                "maxHR": a.get("maxHR"),
                "avgPaceSecPerKm": pace,
                "calories": a.get("calories"),
                "name": a.get("activityName", ""),
                "elevationGain": a.get("elevationGain"),
                "elevationLoss": a.get("elevationLoss"),
            })
        print(f"[{cfg['name']}] 러닝 {len(result)}개 수집 (offset={start})")
        return result
    except Exception as e:
        print(f"[{cfg['name']}] 활동 수집 실패: {e}")
        return []


def fetch_vo2max(client, cfg):
    try:
        today = datetime.date.today().isoformat()
        data = client.get_max_metrics(today)
        if isinstance(data, dict) and data.get("generic"):
            g = data["generic"]
            return [{"date": g.get("calendarDate"), "value": g.get("vo2MaxPreciseValue") or g.get("vo2MaxValue")}]
    except Exception as e:
        print(f"[{cfg['name']}] VO2Max 수집 실패: {e}")
    return None


def fetch_weight(client, cfg):
    try:
        end = datetime.date.today()
        start = end - datetime.timedelta(days=365)
        data = client.get_weigh_ins(start.isoformat(), end.isoformat())
        entries = (data.get("dateWeightList") or data.get("allWeightMetrics") or
                   data.get("weight") or (data if isinstance(data, list) else []))
        result = [
            {"date": e.get("calendarDate") or e.get("date"), "kg": (e.get("weight") or 0) / 1000}
            for e in entries if e.get("weight")
        ]
        return sorted(result, key=lambda x: x["date"], reverse=True) or None
    except Exception as e:
        print(f"[{cfg['name']}] 체중 수집 실패: {e}")
    return None


def fetch_sleep(client, cfg):
    try:
        today = datetime.date.today().isoformat()
        data = client.get_sleep_data(today)
        dto = data.get("dailySleepDTO") or {}
        if dto:
            return [{
                "date": dto.get("calendarDate"),
                "score": (dto.get("sleepScores") or {}).get("overall", {}).get("value"),
                "durationMin": round(dto["sleepTimeSeconds"] / 60) if dto.get("sleepTimeSeconds") else None,
            }]
    except Exception as e:
        print(f"[{cfg['name']}] 수면 수집 실패: {e}")
    return None


def fetch_readiness(client, cfg):
    try:
        today = datetime.date.today().isoformat()
        data = client.get_training_readiness(today)
        items = data if isinstance(data, list) else [data]
        tr = items[0] if items else {}
        if tr.get("score") is not None:
            return {
                "date": tr.get("calendarDate", today),
                "score": tr["score"],
                "level": tr.get("level"),
                "feedbackShort": tr.get("feedbackShort"),
                "sleepScore": tr.get("sleepScore"),
                "recoveryTimeMin": tr.get("recoveryTime"),
            }
    except Exception as e:
        print(f"[{cfg['name']}] 훈련 준비도 수집 실패: {e}")
    return None


def merge_data(existing, fresh):
    if not existing:
        return fresh

    # 기존 활동을 id 기준 dict로 — 새 데이터로 업데이트 (elevation 등 신규 필드 반영)
    existing_by_id = {a["id"]: a for a in existing.get("activities", [])}
    for a in fresh.get("activities", []):
        if a["id"] in existing_by_id:
            existing_by_id[a["id"]].update(a)
        else:
            existing_by_id[a["id"]] = a
    merged_acts = sorted(existing_by_id.values(), key=lambda a: a["date"], reverse=True)[:200]

    existing_w_dates = {w["date"] for w in existing.get("weight") or []}
    merged_weight = sorted(
        (existing.get("weight") or []) + [w for w in (fresh.get("weight") or []) if w["date"] not in existing_w_dates],
        key=lambda w: w["date"], reverse=True
    ) or None

    existing_v2_dates = {v["date"] for v in existing.get("vo2max") or []}
    merged_v2 = sorted(
        (existing.get("vo2max") or []) + [v for v in (fresh.get("vo2max") or []) if v["date"] not in existing_v2_dates],
        key=lambda v: v["date"], reverse=True
    ) or None

    return {**existing, **fresh, "activities": merged_acts, "weight": merged_weight, "vo2max": merged_v2}


def sync_user(user_id, backfill=False):
    cfg = USERS[user_id]
    data_file = cfg["data_file"]

    client = get_client(cfg)

    age = datetime.date.today().year - cfg["birth_year"] if cfg.get("birth_year") else None

    if backfill:
        # 200개까지 페이지 단위로 fetch해서 고도 데이터 백필
        print(f"[{cfg['name']}] 백필 모드 — 최대 200개 활동 고도 데이터 수집 중...")
        all_acts = []
        for offset in range(0, 200, 50):
            batch = fetch_activities(client, cfg, start=offset, limit=50)
            all_acts.extend(batch)
            if len(batch) < 50:
                break
        activities = all_acts
    else:
        activities = fetch_activities(client, cfg)

    fresh = {
        "userId": user_id,
        "name": cfg["name"],
        "birthYear": cfg.get("birth_year"),
        "age": age,
        "mafHR": 180 - age if age else None,
        "targetWeightKg": cfg.get("target_weight_kg"),
        "goalDistanceKm": cfg.get("goal_distance_km"),
        "goalTimeMin": cfg.get("goal_time_min"),
        "activities": activities,
        "vo2max": fetch_vo2max(client, cfg),
        "weight": fetch_weight(client, cfg),
        "sleep": fetch_sleep(client, cfg),
        "trainingReadiness": fetch_readiness(client, cfg),
        "lastSync": datetime.datetime.utcnow().isoformat() + "Z",
    }

    existing = None
    if data_file.exists():
        try:
            existing = json.loads(data_file.read_text())
        except Exception:
            pass

    merged = merge_data(existing, fresh)
    data_file.parent.mkdir(exist_ok=True)
    data_file.write_text(json.dumps(merged, indent=2, ensure_ascii=False))
    print(f"[{cfg['name']}] 저장 완료 → {data_file}")


def main():
    args = sys.argv[1:]
    user_arg = None
    backfill = "--backfill" in args

    if "--user" in args:
        user_arg = args[args.index("--user") + 1]
    elif any(a.startswith("--user=") for a in args):
        user_arg = next(a.split("=", 1)[1] for a in args if a.startswith("--user="))

    users_to_sync = [user_arg] if user_arg else list(USERS.keys())

    for uid in users_to_sync:
        if uid not in USERS:
            print(f"알 수 없는 유저: {uid}")
            continue
        try:
            sync_user(uid, backfill=backfill)
        except Exception as e:
            print(f"[{USERS[uid]['name']}] 동기화 실패: {e}")


if __name__ == "__main__":
    main()
